src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+
+Computes the gradients used in `optimize!`. Default implementation is provided for `VariationalInference{AD}` where `AD` is either `ForwardDiffAD` or `TrackerAD`. This implicitly also gives a default implementation of `optimize!`.
+
+Variance reduction techniques, e.g. control variates, should be implemented in this function.
+"""
+grad!(vo, vi::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ, out, args...) where AD =
+    error("Turing.Variational.grad!: unmanaged variational inference algorithm: $(typeof(alg))")
+
+"""
+    vi(model::Model, alg::VariationalInference)
+    vi(model::Model, alg::VariationalInference, q::VariationalPosterior)
+
+Constructs the variational posterior from the `model` and performs the optimization following the configuration of the given `VariationalInference` instance.
+"""
+vi(model::Model, alg::VariationalInference) = 
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+vi(model::Model, alg::VariationalInference, q::VariationalPosterior) =
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+
+# default implementations
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: ForwardDiffAD}
+    # TODO: this probably slows down executation quite a bit; exists a better way of doing this?
+    f(θ_) = - vo(alg, q, model, θ_, args...)
+
+    chunk_size = getchunksize(alg)
+    # Set chunk size and do ForwardMode.
+    chunk = ForwardDiff.Chunk(min(length(θ), chunk_size))
+    config = ForwardDiff.GradientConfig(f, θ, chunk)
+    ForwardDiff.gradient!(out, f, θ, config)
+end
+
+# TODO: implement for `Tracker`
+# function grad(vo::ELBO, alg::ADVI, q::MeanField, model::Model, f, autodiff::Val{:backward})
+#     vo_tracked, vo_pullback = Tracker.forward()
+# end
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: TrackerAD}
+    θ_tracked = Tracker.param(θ)
+    y = - vo(alg, q, model, θ_tracked, args...)
+    Tracker.back!(y, 1.0)
+
+    DiffResults.value!(out, Tracker.data(y))
+    DiffResults.gradient!(out, Tracker.grad(θ_tracked))
+end
+
+"""
+    optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad())
+
+Iteratively updates parameters by calling `grad!` and using the given `optimizer` to compute the steps.
+"""
+function optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad()) where AD
+    # TODO: this default constructor for `optimizer` is probably dangerous because of the `acc` field being an `IdDict`!!!
+
+    # TODO: should we always assume `samples_per_step` and `max_iters` for all algos?
```

Comment:
It's OK to assume so now. Or maybe we can set them as keyword arguments to `vi(;max_iters=???)`?

Reply:
That's definitively a possible solution. The only problem is that not _all_ variational inference problems will need this, e.g. ELBO and it's gradients exists in closed form. Then it becomes a question of where the boundary is for a parameter to be a keyword argument, as opposed to being attached to the `VariationalInfernece` instance.